File: matlab/iw.m
function [W,pred,iw] = iw(X,Z,y,varargin)
% Implementation of an importance-weighted classifier
%
% References:
% Kernel density estimators.
% Ratio of Gaussians: Shimodaira (2000), Improving predictive inference under covariate shift by weighting the log-likelihood function. JSPI.
% Logistic discrimination: Bickel et al. (2009), Discriminative learning under covariate shift. JMLR.
% Kernel mean matching: Huang et al. (2007), Correcting sample selection bias by unlabeled data. NIPS.
% Nearest-neighbour-based: Loog (2015), Nearest neighbor-based importance weighting. MLSP.
%
% Input:    X       source data (N samples x D features)
%           Z       target data (M samples x D features)
%           y       source labels (N x 1) in {1,...,K}
% Optional:
%           l2      additional l2-regularization parameters (default: 1e-3)
%           iwe     choice of importance weight estimation method (default: 'lr')
%           loss    choice of loss function (default: 'log')
%
% Output:   W       trained linear classifier
%           pred    predictions by trained classifier on target data
%           iw      estimated importance weights for source samples
%
% Copyright: Wouter M. Kouw
% Last update: 19-12-2017

% Add dependencies to path
addpath(genpath('util'));
addpath(genpath('minFunc'));

% Check for solver
if isempty(which('minFunc')); error('Can not find minFunc'); end
options.DerivativeCheck = 'off';
options.Method = 'lbfgs';
options.Display = 'final';

% Parse optionals
p = inputParser;
addOptional(p, 'l2', 1e-3);
addOptional(p, 'iwe', 'log');
addOptional(p, 'loss', 'log');
parse(p, varargin{:});

% Data shape
[N,D] = size(X);
[M,E] = size(Z);
labels = unique(y);
K = length(labels);

% Check if dimensionalities are the same
if D~=E; error('Data dimensionalities not the same in both domains'); end

% Check if labels are in [1,...,K]
if ~isempty(setdiff(labels,1:K)); error('Labels should be in [1,...,K]'); end

% Check for bias augmentation
if ~all(X(:,end)==1) && ~all(Z(:,end)==1)
    X = [X ones(N,1)];
    Z = [Z ones(M,1)];
    D = D+1;
end

% Estimate importance weights
switch lower(p.Results.iwe)
    case {'lr', 'log'}
        iw = iwe_lr(X,Z, p.Results.l2);
    case {'rg', 'gauss'}
        iw = iwe_rg(X(:,1:end-1),Z(:,1:end-1), p.Results.l2);
    case {'kd', 'kde'}
        iw = iwe_kd(X(:,1:end-1),Z(:,1:end-1), p.Results.l2);
    case 'kmm'
        iw = iwe_kmm(X(:,1:end-1),Z(:,1:end-1), 1,'rbf');
    case 'nn'
        iw = iwe_nn(X(:,1:end-1),Z(:,1:end-1));
    otherwise
        print('Importance-weight estimator unknown')
end

switch p.Results.loss
    case {'ls','qd'}
        % Map y to one-hot encoding in {-1,+1} (= one-v-rest)
        Y = -ones(N,K);
        for n = 1:N
            Y(n,y(n)) = 1;
        end

        % Least-squares solution to importance-weighted quadratic loss
        W = (X'*diag(iw)*X + p.Results.l2*eye(D))\(X'*diag(iw)*Y);

    case {'lr','log'}
        % Minimize loss
        W = minFunc(@mwlr_grad, zeros(D*K,1), options, X, y, iw, p.Results.l2);
        W = reshape(W, [D,K]);

    otherwise
        error('Loss function not implemented');
end

% Predict target labels
[~,pred] = max(Z*W, [], 2);

end

function [L, dL] = mwlr_grad(W,X,y,iw, l2)
% Multi-class importance-weighted logistic regression gradient
% Function expects bias-augmented X and y in [1,...,K]

% Shapes
[~,D] = size(X);
labels = unique(y);
K = length(labels);
W = reshape(W, [D,K]);

% Numerical stability
XW = X*W;
max_const = max(XW, [], 2);
XW = bsxfun(@minus, XW, max_const);

% Point-wise weighted negative log-likelihood
L = -iw.*sum(X.*W(:,y)',2) + iw.*log(sum(exp(XW),2)) + iw.*max_const;

% Add l2-regularizer
L = sum(L,1) + l2 .* sum(W(:).^2);

% Only compute gradient if requested
if nargout > 1

    % Gradient with respect to B_k
    dL = zeros(D,K);
    for k = 1:K
        dL(:,k) = sum( -bsxfun(@times,iw.*(y==labels(k)),X)' + ...
            bsxfun(@times,iw.*exp(XW(:,k))./sum(exp(XW),2),X)',2);
    end

	% Add l2-regularizer
	dL = dL(:) + 2*l2.*W(:);

end

end

File: python/iw.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import numpy as np
import scipy.stats as st
from scipy.spatial.distance import cdist
import sklearn as sk
from sklearn.SVM import LinearSVC
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.model_selection import cross_val_predict
from os.path import basename

from util import is_pos_def


class ImportanceWeightedClassifier(object):
    """
    Class of importance-weighted classifiers.

    Methods contain different importance-weight estimators and different loss
    functions.
    """

    def __init__(self, iwe='lr', l2=1.0, smoothing=True, clip=-1,
                 loss='logistic'):
        """
        Select a particular type of importance-weighted classifier.

        INPUT   (1) str 'iwe': importance-weight estimator (def:'lr')
                (2) float 'l2': l2-regularization parameter value (def:0.01)
                (3) boolean 'smoothing': whether to apply Laplace smoothing to
                    the nearest-neighbour importance-weight estimator
                    (def: True)
                (4) float 'clip': maximum allowable importance-weight value; if
                    set to -1, then the weights are not clipped (def:-1)
                (5) str 'loss': loss function for weighted classifier, options:
                    'logistic', 'quadratic', 'hinge' (def: 'logistic')
        OUTPUT  (1) array 'preds': predictions on given target data (M samples
                    by 1)
                (2) array 'theta': linear classifier parameters (D features by
                    K classes)
        """
        self.iwe = iwe
        self.l2 = l2
        self.smoothing = smoothing
        self.clip = clip
        self.loss = loss

    def iwe_gauss(self, X, Z):
        """
        Estimate importance weights based on a ratio of Gaussian distributions.

        INPUT   (1) array 'X': source data (N samples by D features)
                (2) array 'Z': target data (M samples by D features)
        OUTPUT  (1) array: importance weights (N samples by 1)
        """
        # Sample means in each domain
        mu_X = np.mean(X, axis=0)
        mu_Z = np.mean(Z, axis=0)

        # Sample covariances
        Si_X = np.cov(X)
        Si_Z = np.cov(Z)

        # Check for positive-definiteness of covariance matrices
        if not (is_pos_def(Si_X) or is_pos_def(Si_Z)):
            print('Warning: covariate matrices not PSD.')

            regct = -6
            while not (is_pos_def(Si_X) or is_pos_def(Si_Z)):
                print('Adding regularization: ' + str(1**regct))

                # Add regularization
                Si_X += 1**regct
                Si_Z += 1**regct

                # Increment regularization counter
                regct += 1

        # Compute probability of X under each domain
        pT = st.multivariate_normal(X, mu_Z, Si_Z)
        pS = st.multivariate_normal(X, mu_X, Si_X)

        # Return the ratio of Gaussians
        return pT / pX

    def iwe_lr(self, X, Z):
        """
        Estimate importance weights based on logistic regression.

        INPUT   (1) array 'X': source data (N samples by D features)
                (2) array 'Z': target data (M samples by D features)
        OUTPUT  (1) array: importance weights (N samples by 1)
        """
        # Data shapes
        N, DX = X.shape
        M, DZ = Z.shape

        # Assert equivalent dimensionalities
        if DX != DZ:
            raise AssertionError

        # Make domain-label variable
        y = np.concatenate((np.zeros((N, 1)),
                            np.ones((M, 1))), axis=0)

        # Concatenate data
        XZ = np.concatenate((X, Z), axis=0)

        # Call a logistic regressor
        lr = LogisticRegression(C=self.l2)

        # Predict probability of belonging to target using cross-validation
        preds = cross_val_predict(lr, XZ, y[:, 0])

        # Return predictions for source samples
        return preds[:N]

    def iwe_nn(self, X, Z):
        """
        Estimate importance weights based on nearest-neighbours.

        INPUT   (1) array 'X': source data (N samples by D features)
                (2) array 'Z': target data (M samples by D features)
        OUTPUT  (1) array: importance weights (N samples by 1)
        """
        # Number of source samples
        N = X.shape[0]

        # Compute Euclidean distance between samples
        d = cdist(X, Z, metric='euclidean')

        # Count target samples within each source Voronoi cell
        ix = np.argmin(d, axis=1)
        iw = hist(ix, np.arange(N))

        # Laplace smoothing
        if self.smoothing:
            iw = (iw + 1) / (N + 1)

        # Weight clipping
        if self.clip > 0:
            iw = min(self.clip, max(0, iw))

        # Return weights
        return iw

    def fit(self, X, y, Z):
        """
        Fit/train an importance-weighted classifier.

        INPUT   (1) array 'X': source data (N samples by D features)
                (2) array 'y': source labels (N samples by 1)
                (3) array 'Z': target data (M samples by D features)
                (4) str 'iwe': importance weight estimator,
                    options: 'lr', 'nn', 'rg', 'kmm', 'kde' (def: 'lr')
        OUTPUT  (1) array 'theta': trained classifier parameters
                (2) array 'preds': predictions of trained classifier on given
                    target samples
        """
        # Find importance-weights
        if self.iwe == 'lr':
            w = self.iwe_lr(X, Z)
        elif self.iwe == 'gauss':
            w = self.iwe_lr(X, Z)
        elif self.iwe == 'nn':
            w = self.iwe_nn(X, Z)
        elif self.iwe == 'kde':
            w = self.iwe_kde(X, Z)
        elif self.iwe == 'kmm':
            w = self.iwe_kmm(X, Z)

        # Train a weighted classifier
        if self.loss == 'logistic':
            # Logistic regression model with sample weights
            iwclf = LogisticRegression().fit(X, y, w)

        elif self.loss == 'quadratic':
            # Least-squares model with sample weights
            iwclf = LinearRegression().fit(X, y, w)

        elif self.loss == 'hinge':
            # Linear support vector machine with sample weights
            iwclf = LinearSVC().fit(X, y, w)

        # Get trained classifier parameters
        theta = iwclf.get_params()

        # Make predictions on given target data
        preds = iwclf.predict(Z)

        # Return classifier parameters and predictions
        return theta, preds
